config/filter.d/dropbear.conf

```diff
@@ -23,7 +23,7 @@ before = common.conf
 
 _daemon = dropbear
 
-prefregex = ^%(__prefix_line)s<F-CONTENT>(?:[Ll]ogin|[Bb]ad|[Ee]xit).+</F-CONTENT>$
+prefregex = ^%(__prefix_line)s(\[\d+\] [A-Z][a-z]+ \d\d \d\d:\d\d:\d\d )?<F-CONTENT>(?:[Ll]ogin|[Bb]ad|[Ee]xit).+</F-CONTENT>$
```

Comment:
The issue with that additional timestamp - it looks like a default date-time format, so that for example it may depend on current locale.
Basically it can be any of 
https://github.com/fail2ban/fail2ban/blob/8170e9fe75fd2c2c4c51a1d9972b683401cddccb/fail2ban/server/datedetector.py#L126-L157
Also `[A-Z][a-z]+ \d\d` would not match `Mär 10` (umlaut) or `Jul  1` (space and single digit)...
There is no good RE to cover it (no special chars or separator after the timestamp), and catch-all like `.*` is too ugly.
I'd suggest something like `(?:\[\d+\](?:\s+\w{2,3}(?:(?:\s+|[:\-\./])\w{1,3}){0,2})?(?:(?:\s+|[:\-\./])\d+){0,4}\d )?` but it also looks a bit like a workaround to me. Although since all date formats mostly ends with a digit (and followed by space), I think it may be acceptable and `\d ` can be considered as an "anchor" for the timestamp, before the content of message begins.

Reply:
Dropbear is no localised. However, of course it could use a system library to generate the timestamp, which is. I just applied German locale on a test system, and added `LC_ALL=de_DE.UTF-8` to the system's and unit's environment, and added `ExecStartPre=/bin/dash -c 'date; locale'` to assure it is all applied:
```
Mär 01 13:38:19 VM-Bookworm systemd[1]: Starting dropbear.service - Lightweight SSH server...
Mär 01 13:38:19 VM-Bookworm dash[848]: Fr 1. Mär 13:38:19 CET 2024
Mär 01 13:38:19 VM-Bookworm dash[849]: LANG=de_DE.UTF-8
Mär 01 13:38:19 VM-Bookworm dash[849]: LANGUAGE=
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_CTYPE="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_NUMERIC="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_TIME="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_COLLATE="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_MONETARY="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_MESSAGES="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_PAPER="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_NAME="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_ADDRESS="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_TELEPHONE="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_MEASUREMENT="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_IDENTIFICATION="de_DE.UTF-8"
Mär 01 13:38:19 VM-Bookworm dash[849]: LC_ALL=de_DE.UTF-8
Mär 01 13:38:19 VM-Bookworm systemd[1]: Started dropbear.service - Lightweight SSH server.
Mär 01 13:38:19 VM-Bookworm dropbear[850]: [850] Mar 01 13:38:19 Failed loading /etc/dropbear/dropbear_dss_host_key
Mär 01 13:38:19 VM-Bookworm dropbear[850]: [850] Mar 01 13:38:19 Not backgrounding
```
As can be seen, despite German locale being applied, `journalctl` and `date` (within the unit) returning the problematic timestamp, Dropbear sticks to the expected one.

But otherwise, would `{^LN-BEG}` or `%(datepattern)s` (from `common.conf`) cover all cases, to be failsafe, in case builds on other distros, or if the build time locale(config have an effect?

Reply:
> As can be seen, despite German locale being applied, journalctl and date (within the unit) returning the problematic timestamp

In reality it doesn't really (see below). It's just a text representation... in journal entry it is stored in numeric (binary) form.

> But otherwise, wouldn't `{DATE}` cover all cases, to be failsafe...

No, it would not.
1. `datepattern` only used by file-related backends, not for systemd backend, where the time will be obtained directly from journal entry (in form of datetime object without to parse it at all).
2. therefore the filter will not search the date at all (neither the first, what is simply representation in text-form, nor the second, logged by the service logger in text form) and therefore it'd not cut it out from message before `prefregex` or `failregex` are applied. No date expected, no extraction happens.

Reply:
> In reality it doesn't really (see below). It's just a text representation... in journal entry it is stored in numeric (binary) form.

What I mean is that in login shell (`/etc/default/locale`), as well as the systemd unit's context, the German locale is applied, but Dropbear still logs in international date format.

> No, it would not.

I see. So `{DATE}` is not a generic date pattern for free usage, but one to match the log timestamp so that Fail2Ban can apply its `findtime` rules and such? _Since there is already the `datedetector` with a regex template, it would be actually nice to make it available for filters, independent of whether/how Fail2Ban extracts the timestamp for its own purpose. When daemons do not provide a particular journal logging mode, but STDOUT for logging to foreground console/fd in the first place, doubled timestamps in journal are quite common. But this is a different topic 🙂._

Regarding the question whether Dropbear logs its timestamps in different format depending on host OS, build environment/config or such, before guessing around or applying a possibly unnecessary complex regex, maybe @mkj can answer this faster than I am able to check and understand the code?

Reply:
Dropbear is logging with `strftime` `"%b %d %H:%M:%S`.

https://github.com/mkj/dropbear/blob/2674736bce847bc51a30a07c191562047659ac1f/src/svr-session.c#L338-L341

Reply:
Better would be surely not to log `[825] Jul 10 23:57:29` at all (for instance by something like `-F -E --log-msg-only` or `--journal`), since it simply doesn't make sense.

Reply:
Okay great, so we know the two digit day and time format for sure. The only thing is that `%b` should use the current locale. Not sure why it does not in my tests, or it is set/overridden elsewhere. To be failsafe, we could use `\w+` or `\w{2,3}` instead `[A-Z][a-z]+`?

A journal logging mode for Dropbear would be great, in combination with an upstream systemd service, for completion. However, we would need to support the additional timestamp in any case, so it does not matter for this PR.

Reply:
> To be failsafe, we could use `\w+` or `\w{2,3}` instead `[A-Z][a-z]+`?

`\w{2,3}` is fully enough. To simplify the whole thing - `(?:\[\d+\] \w{2,3} [\d:\s]+)?`. Note `(?:...)` for non-capturing group.

> However, we would need to support the additional timestamp in any case, so it does not matter for this PR.

No, we would not... the logger can simply avoid to write this part at all in that cases.
The journal entry need only contain the message, everything else is made by the wrapper stderr to journal automatically.
Here is for instance the code how it's done in fail2ban (for better understanding):
https://github.com/fail2ban/fail2ban/blob/8170e9fe75fd2c2c4c51a1d9972b683401cddccb/fail2ban/server/server.py#L739-L743

For this particular case it'd be enough something like that in https://github.com/mkj/dropbear/blob/2674736bce847bc51a30a07c191562047659ac1f/src/svr-session.c#L346:
```diff
+		if (!opts.logmsgonly) {
-		fprintf(stderr, "[%d] %s %s\n", getpid(), datestr, printbuf);
+			fprintf(stderr, "[%d] %s %s\n", getpid(), datestr, printbuf);
+		} else {
+			fprintf(stderr, "%s\n", printbuf);
+		}
```

Reply:
> To simplify the whole thing - `(?:\[\d+\] \w{2,3} [\d:\s]+)?`. Note `(?:...)` for non-capturing group.

I saw the syntax by times, but looked it up just when you posted the complex regex above. Makes sense if there is no `\1`-like reference done, of if we do not know parent or child pattern groups 👍. Let's go with that one.

> No, we would not... the logger can simply avoid to write this part at all in that cases.

What I mean is that, even when Dropbear implemented this logging mode now, it would take years until it trickled down into stable releases of major distros. And even then, we cannot be sure which Dropbear version is really used, and whether this logging mode is really used, even if available. I do not know the policy about how compatible upstream Fail2Ban filters shall be, but I personally would just keep supporting all possible cases, even if this logging option gets implemented.